src/pcp/types.rs

```diff
@@ -259,124 +216,101 @@ impl ValueParam for SumValueParam {
 /// The histogram type. Each measurement is a non-negative integer and the aggregate is a histogram
 /// approximating the distribution of the measurements.
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub struct Histogram<F> {
-    data: Vec<F>,
+pub struct Histogram<F: FieldElement> {
+    buckets: Vec<F::Integer>,
     range_checker: Vec<F>,
-    sum_check_share: F,
 }
 
 impl<F: FieldElement> Histogram<F> {
-    /// Constructs a new histogram input. The values of `buckets` must be strictly increasing.
-    pub fn new(measurement: u64, buckets: &[u64]) -> Result<Self, PcpError> {
-        let mut data = vec![F::zero(); buckets.len() + 1];
+    /// Return a new [`Histogram`] type with the given buckets.
+    pub fn new(buckets: Vec<F::Integer>) -> Self {
+        Self {
+            buckets,
+            range_checker: poly_range_check(0, 2),
+        }
+    }
+}
+
+impl<F: FieldElement> Type for Histogram<F> {
+    type Measurement = F::Integer;
+    type Field = F;
+
+    fn encode(&self, measurement: &F::Integer) -> Result<Vec<F>, PcpError> {
+        let mut data = vec![F::zero(); self.buckets.len() + 1];
 
-        if buckets.len() >= u32::MAX as usize {
-            return Err(PcpError::Value(
+        // XXX Move this check to constructor
+        if self.buckets.len() >= u32::MAX as usize {
+            return Err(PcpError::Encode(
                 "invalid buckets: number of buckets exceeds maximum permitted".to_string(),
             ));
         }
 
-        if !buckets.is_empty() {
-            for i in 0..buckets.len() - 1 {
-                if buckets[i + 1] <= buckets[i] {
-                    return Err(PcpError::Value(
+        // XXX Move this check to constructor
```

Comment:
Any reason not to do this now, and make `Histogram::new` fallible? Same goes for `Sum::new`.

Reply:
The reason to do this in the constructor is that we only need to validate that `self.buckets` is well-formed once. As it stands, we end up re-validating this every time we call `encode`.